### DISCUSSION TITLE
On demand atom based on ID

### DISCUSSION THREAD

**Discussion by tpatalas** [pmndrs/jotai#1337] [Q&A] [answered]
I first should say enjoying using Jotai very much. Jotai makes code written with Recoil simpler and cleaner, and I love the facts that Jotai is based on weakMap, so we do not worry about Memory leaks. Plus, It offers splitAtom, selectAtom, and so on to improve the performance. 

One thing I couldn't find it intuitive is this:

I can open and close the each modals based on the given ID. With below code I can reuse everywhere as long as id is provided (or undefined also works). 
```ts
// Recoil:
export const atomTodoModalOpen = atomFamily({
  key: `atomTodoModalOpen`,
  default: false,
});
```

I tried atomFamily from Jotai, but couldn't really make it work plus I do not want to use atomFamily from Jotai since it is based on Map instead of weakMap that I have to care for memory leak again like I did in Recoil. 

So I did this:
```ts
// Jotai:
export const atomModals = atom<modals[]>([]);
export const atomItemModal = atom({} as modals);
export const atomModalOpen = (_id) => {
  const anAtom = atom(
    (get) => {
      const itemModal = get(atomModals).find((modal) => modal._id === _id);
      return itemModal?.initial === undefined ? false : itemModal.initial;
    },
    (get, set) => {
      set(atomItemModal, { ...get(atomItemModal), _id, initial: false });
      set(atomModals, [...get(atomModals), get(atomItemModal)]);
    }
  );
  anAtom.onMount = (setAtom) => setAtom();
  return anAtom;
};

```
It simply creates the object if given ID is not found. 

Although It works as expected, It is not so intuitive as you can compare those two codes. So I was wondering if there is better way I can do it in Jotai way?

**Comment by dai-shi** [comment, marked as answer]
If you have finite number of `id`s, I think `atomFamily` is just easy.
```js
const todoModalOpenAtomFamily = atomFamily(() => atom(false))
```

For the infinite `id`s, I would use atoms-in-atom pattern, but it would still require to remove an item of unused `id` from a list. So, essentially, it's the same as `todoModalOpenAtomFamily.remove(id)`.

If we use atoms-in-atom, we should think like eliminating `id`s at all. Well, the fact is `id` is not garbage collectable anyway...

In summary:
1. if you can eliminate `id`, and make atoms instead, things are garbage collectable.
2. otherwise, as `id` is not garbage collectable, just use `atomFamily` and `.remove()`.

**Reply by tpatalas** [reply]
Okay after seeing your explanation. I found that I completely misunderstood the atomFamily of Jotai. I thought I must pass the parameter, but it accepts as it is shown on your code. So It is basically same as Recoil's, and what I did above was completely ignorable. Great!  

I guess I gotta read Jotai's docs better. Thank you for your input @dai-shi. It really helps others to understand Jotai better.

**Reply by dai-shi** [reply]
Good hear that. Please also check out #1323.

Especially, `atomFamily` is as simple as this: https://github.com/pmndrs/jotai/discussions/1323#discussioncomment-3303035
```js
// simplified code
const atomFamily = (initializeAtom) => {
  const cache = new Map();
  return (param) => {
    if (!cache.has(param)) {
      cache.set(param, initializeAtom(param));
    }
    return cache.get(param);
  };
};
```
